crates/clvm-traits/docs/derive_macros.md

```diff
@@ -12,102 +12,183 @@ Pick whichever representation fits your use-case the best.
-arguments are quoted and terminated with `1`, which is how partial application is implemented in CLVM.
+### Optional Value
 
-You can read more about currying on the [Chia blockchain documentation](https://docs.chia.net/guides/chialisp-currying).
+You can specify a field as optional directly, which will be set to `None` if it's not present:
+
+```rust
+use clvmr::Allocator;
+use clvm_traits::{ToClvm, FromClvm};
+
+#[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(list)]
+struct Person {
+    name: String,
+    #[clvm(optional)]
+    email: Option<String>,
+}
+
+let person = Person {
+    name: "Bob".to_string(),
+    email: Some("bob@example.com".to_string()),
+};
+
+let a = &mut Allocator::new();
+let ptr = person.to_clvm(a).unwrap();
+assert_eq!(Person::from_clvm(a, ptr).unwrap(), person);
+```
 
-Note that the following code is for example purposes only and is not indicative of how to create a secure program.
-Using a password like shown in this example is an insecure method of locking coins, but it's effective for learning.
+### Default Value
+
+You can also specify the default value manually. The field will not be serialized if it matches the default value:
 
 ```rust
 use clvmr::Allocator;
 use clvm_traits::{ToClvm, FromClvm};
 
 #[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
-#[clvm(curry)]
-struct PasswordArgs {
-    password: String,
+#[clvm(list)]
+struct Person {
+    name: String,
+    #[clvm(default = 18)]
+    age: u8,
 }
 
-let args = PasswordArgs {
-    password: "Hello".into(),
+let person = Person {
+    name: "Bob".to_string(),
+    age: 24,
 };
 
 let a = &mut Allocator::new();
-let ptr = args.to_clvm(a).unwrap();
-assert_eq!(PasswordArgs::from_clvm(a, ptr).unwrap(), args);
+let ptr = person.to_clvm(a).unwrap();
+assert_eq!(Person::from_clvm(a, ptr).unwrap(), person);
+```
+
+## Consume the Rest
+
+You can consume the rest of the list items (or curried arguments, if using the `curry` representation) by using `#[clvm(rest)]`.
+This is useful for types which are represented compactly, without a nil terminator. Or for extending a list of arguments with another.
+You can also use it if you want to lazily parse the rest later.
+
+Here's a simple example of a compact representation:
+
+```rust
+use clvmr::Allocator;
+use clvm_traits::{ToClvm, FromClvm};
+
+#[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(list)]
+struct Point {
+    x: i32,
+    #[clvm(rest)]
+    y: i32,
+}
+
+let point = Point {
+    x: 5,
+    y: 2,
+};
+
+let a = &mut Allocator::new();
+let ptr = point.to_clvm(a).unwrap();
+assert_eq!(Point::from_clvm(a, ptr).unwrap(), point);
+```
+
+And here's an example of lazily parsing the rest later:
+
+```rust
+use clvmr::{Allocator, NodePtr};
+use clvm_traits::{ToClvm, FromClvm};
+
+#[derive(Debug, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(list)]
+struct Items<T> {
+    first: String,
+    #[clvm(rest)]
+    rest: T,
```

Comment:
In this case, `rest` is `NodePtr` in the usage, you just can't put `NodePtr` directly inside of a `ToClvm`/`FromClvm` struct since it implements over the generic node type `Node`